packages/asset-swapper/src/network/samplers/balancer.ts

```diff
@@ -0,0 +1,250 @@
+    ) {
+        super(_cache);
+    }
+
+    protected async _fetchPoolsForPairAsync(takerToken: string, makerToken: string): Promise<Pool[]> {
+        try {
+            const poolData = (await getCreamPoolsWithTokens(takerToken, makerToken)).pools;
+            // Sort by maker token balance (descending)
+            const pools = parseCreamPoolData(poolData, takerToken, makerToken).sort((a, b) =>
+                b.balanceOut.minus(a.balanceOut).toNumber(),
+            );
+            return pools.slice(0, this.maxPoolsFetched);
+        } catch (err) {
+            return [];
+        }
+    }
+}
+
+type SellContract = ERC20BridgeSamplerContract;
+type BuyContract = ERC20BridgeSamplerContract;
+type SellContractSellFunction = SellContract['sampleSellsFromBalancer'];
+type BuyContractBuyFunction = BuyContract['sampleBuysFromBalancer'];
+type SellContractSellFunctionParameters = Parameters<SellContractSellFunction>;
+type BuyContractBuyFunctionParameters = Parameters<BuyContractBuyFunction>;
+type SellContractSellFunctionReturn = UnwrapContractFunctionReturnType<ReturnType<SellContractSellFunction>>;
+type BuyContractBuyFunctionReturn = UnwrapContractFunctionReturnType<ReturnType<BuyContractBuyFunction>>;
+type SamplerSellEthCall = SamplerEthCall<BalancerFillData,SellContractSellFunctionParameters,SellContractSellFunctionReturn>;
+type SamplerBuyEthCall = SamplerEthCall<BalancerFillData,BuyContractBuyFunctionParameters,BuyContractBuyFunctionReturn>;
+
+export class BalancerSampler extends
+    OnChainSourceSampler<
+        SellContract,
+        BuyContract,
+        SellContractSellFunction,
+        BuyContractBuyFunction,
+        BalancerFillData
+    >
+{
+    public static async createAsync(chain: Chain, fork: ERC20BridgeSource): Promise<BalancerSampler> {
+        if (chain.chainId !== ChainId.Mainnet) {
+            throw new Error(`Balancer forks are only available on mainnet`);
+        }
+        let cache: PoolsCache;
+        switch (fork) {
+            case ERC20BridgeSource.Balancer:
+                cache = new BalancerPoolsCache();
+                break;
+            case ERC20BridgeSource.Cream:
+                cache = new CreamPoolsCache();
+                break;
+            default:
+                throw new Error(`Invalid Balancer fork: ${fork}`);
+        }
+        return new BalancerSampler(chain, fork, cache);
+    }
+
+    protected constructor(chain: Chain, public readonly fork: ERC20BridgeSource, private readonly _cache: PoolsCache) {
+        super({
+            chain,
+            sellSamplerContractArtifactName: 'ERC20BridgeSampler',
+            buySamplerContractArtifactName: 'ERC20BridgeSampler',
+            sellSamplerContractType: ERC20BridgeSamplerContract,
+            buySamplerContractType: ERC20BridgeSamplerContract,
+            sellContractSellFunctionName: 'sampleSellsFromBalancer',
+            buyContractBuyFunctionName: 'sampleBuysFromBalancer',
```

Comment:
If anyone can figure out how to infer these values from the prior `type` boilerplate, that'd be cool.